File: python/trainer.py
import os
import time
import numpy as np
import torch
import torch.nn as nn
from typing import Callable, Optional
from .model import Snowman
from .config import SnowballConfig
from .interfaces import DataStream
from .flops import flop_single
from .db import TrainLogger


class SnowballTrainer:
    def __init__(self, model: Snowman, config: SnowballConfig,
                 train_data: DataStream, eval_data: Optional[DataStream] = None,
                 tokens_per_step: Optional[int] = None,
                 on_step: Optional[Callable] = None,
                 on_eval: Optional[Callable] = None,
                 on_phase: Optional[Callable] = None):
        config.validate()
        self.model = model.to(config.device)
        self.cfg = config
        self.train_data = train_data
        self.eval_data = eval_data
        self.tps = tokens_per_step
        self.on_step = on_step
        self.on_eval = on_eval
        self.on_phase = on_phase

        self.P_emb = model.embed_params
        self.B_block = model.block_params
        self.P_head = model.readout_params

        self.logger = None
        if config.db_path:
            self.logger = TrainLogger(config.db_path, config)

    def _steps_per_phase(self, flop_budget: float) -> int:
        # equal steps per phase: budget = S * tps * Σ flop_single(k)
        total_cost = sum(
            flop_single(k, self.P_emb, self.B_block, self.P_head)
            for k in range(1, self.cfg.k_max + 1)
        )
        return int(flop_budget / (self.tps * total_cost))

    def _cosine_lr(self, phase_step: int, phase_total: int) -> float:
        return self.cfg.lr * 0.5 * (1 + np.cos(np.pi * phase_step / phase_total))

    def _save_checkpoint(self, step, K, opt, flops_cum, tag=''):
        cfg = self.cfg
        if not cfg.checkpoint_dir:
            return None
        os.makedirs(cfg.checkpoint_dir, exist_ok=True)
        name = f'step{step}_K{K}{tag}.pt'
        path = os.path.join(cfg.checkpoint_dir, name)
        torch.save({
            'step': step, 'K': K, 'flops_cum': flops_cum,
            'model': self.model.state_dict(),
            'optimizer': opt.state_dict(),
        }, path)
        if self.logger:
            self.logger.log_checkpoint(cfg.run_name, step, K, path)
            self.logger.commit()
        return path

    @torch.no_grad()
    def evaluate(self, K: int) -> float:
        self.model.eval()
        total = 0.0
        for _ in range(self.cfg.eval_batches):
            x, y = self.eval_data.get_batch()
            logits = self.model(x, K)
            total += self.cfg.loss_fn(logits, y).item()
        self.model.train()
        return total / self.cfg.eval_batches

    def train(self, flop_budget: Optional[float] = None) -> dict:
        torch.manual_seed(self.cfg.seed)
        cfg = self.cfg
        run = cfg.run_name
        use_flop = flop_budget is not None and self.tps is not None

        model = self.model
        model.train()
        opt = cfg.make_optimizer(model.parameters())

        t0 = time.time()
        flops_cum = 0.0
        best_loss = float('inf')
        step = 0
        current_K = cfg.start_k
        phase_step = 0

        if use_flop:
            phase_total = self._steps_per_phase(flop_budget)
        else:
            if cfg.total_steps is None:
                raise ValueError("need flop_budget or total_steps")
            phase_total = cfg.total_steps

        if self.logger:
            self.logger.log_phase(run, current_K, step, flops_cum)
            self.logger.commit()

        while True:
            if phase_step >= phase_total:
                if current_K >= cfg.k_max:
                    break
                old_K = current_K
                current_K += 1
                phase_step = 0
                opt = cfg.make_optimizer(model.parameters())
                if cfg.checkpoint_on_phase and cfg.checkpoint_dir:
                    self._save_checkpoint(step, current_K, opt, flops_cum, '_phase')
                if self.logger:
                    self.logger.log_phase(run, current_K, step, flops_cum)
                    self.logger.commit()
                if self.on_phase:
                    self.on_phase(old_K, current_K, step, flops_cum)

            lr = self._cosine_lr(phase_step, phase_total)
            for pg in opt.param_groups:
                pg['lr'] = lr

            x, y = self.train_data.get_batch()
            opt.zero_grad()
            # random depth sampling: k ~ U(1..current_K)
            # E[grad] ∝ multi-depth grad; AdamW m/√v normalises the scale
            k = torch.randint(1, current_K + 1, (1,)).item()
            ce = model.forward_single(x, k, y, cfg.loss_fn)
            nn.utils.clip_grad_norm_(model.parameters(), cfg.grad_clip)
            opt.step()

            if self.tps is not None:
                fpt = flop_single(k, self.P_emb, self.B_block, self.P_head)
                flops_cum += fpt * self.tps

            if step % cfg.log_interval == 0:
                gates = model.gate_values[:current_K]
                if self.logger:
                    self.logger.log_step(run, step, current_K, ce,
                                         lr, flops_cum, gates, [ce])
                    self.logger.commit()
                if self.on_step:
                    self.on_step(step, current_K, ce, lr, flops_cum)

            if (step + 1) % cfg.eval_interval == 0 and self.eval_data is not None:
                val_loss = self.evaluate(current_K)
                if val_loss < best_loss:
                    best_loss = val_loss
                    if cfg.checkpoint_best and cfg.checkpoint_dir:
                        self._save_checkpoint(step + 1, current_K, opt, flops_cum, '_best')
                if self.logger:
                    self.logger.log_eval(run, step + 1, current_K, val_loss, flops_cum)
                    self.logger.commit()
                if self.on_eval:
                    self.on_eval(step + 1, current_K, val_loss, flops_cum)

            if cfg.checkpoint_interval and cfg.checkpoint_dir:
                if (step + 1) % cfg.checkpoint_interval == 0:
                    self._save_checkpoint(step + 1, current_K, opt, flops_cum)

            step += 1
            phase_step += 1

        elapsed = time.time() - t0

        final_loss = None
        if self.eval_data is not None:
            final_loss = self.evaluate(cfg.k_max)
            if final_loss < best_loss:
                best_loss = final_loss

        result = {
            'best_loss': best_loss,
            'final_loss': final_loss if final_loss is not None else ce,
            'total_steps': step,
            'elapsed': elapsed,
            'total_flops': flops_cum,
            'params': model.param_count,
            'block_params': model.block_params,
        }

        if self.logger:
            for k, v in result.items():
                self.logger.log_summary(run, k, v)
            self.logger.commit()
            self.logger.close()

        return result

File: rust/src/trainer.rs
use std::time::Instant;

use burn::prelude::*;
// AutodiffModule is used implicitly via Module derive
use burn::optim::{AdamWConfig, GradientsParams, Optimizer};
use burn::grad_clipping::GradientClippingConfig;

use crate::config::SnowballConfig;
use crate::db::TrainLogger;
use crate::flops::flop_single;
use crate::model::Snowman;

pub struct TrainResult {
    pub best_loss: f64,
    pub final_loss: f64,
    pub total_steps: usize,
    pub elapsed: f64,
    pub total_flops: f64,
    pub params: usize,
    pub block_params: usize,
}

pub fn train<B>(
    mut model: Snowman<B>,
    cfg: &SnowballConfig,
    device: &B::Device,
    tokens_per_step: usize,
    flop_budget: Option<f64>,
    mut train_data: impl FnMut(&B::Device) -> (Tensor<B, 2, Int>, Tensor<B, 2, Int>),
    mut eval_data: Option<&mut dyn FnMut(&B::Device) -> (Tensor<B, 2, Int>, Tensor<B, 2, Int>)>,
    mut on_step: Option<&mut dyn FnMut(usize, usize, f64, f64, f64)>,
    mut on_eval: Option<&mut dyn FnMut(usize, usize, f64, f64)>,
    mut on_phase: Option<&mut dyn FnMut(usize, usize, usize, f64)>,
) -> (Snowman<B>, TrainResult)
where
    B: burn::tensor::backend::AutodiffBackend,
{
    B::seed(device, cfg.seed);

    let p_emb = model.embed_param_count();
    let b_block = model.block_param_count();
    let p_head = model.readout_param_count();

    let mut optim = AdamWConfig::new()
        .with_weight_decay(cfg.weight_decay as f32)
        .with_grad_clipping(Some(GradientClippingConfig::Norm(cfg.grad_clip as f32)))
        .init::<B, Snowman<B>>();

    let logger = cfg.db_path.as_ref().map(|p| TrainLogger::open(p));

    let phase_total = if let Some(budget) = flop_budget {
        let total_cost: usize = (1..=cfg.k_max)
            .map(|k| flop_single(k, p_emb, b_block, p_head))
            .sum();
        (budget / (tokens_per_step as f64 * total_cost as f64)) as usize
    } else {
        cfg.total_steps.expect("need flop_budget or total_steps")
    };

    let t0 = Instant::now();
    let mut flops_cum: f64 = 0.0;
    let mut best_loss = f64::INFINITY;
    let mut step: usize = 0;
    let mut current_k: usize = cfg.start_k;
    let mut phase_step: usize = 0;
    let mut last_ce: f64 = 0.0;

    if let Some(ref lg) = logger {
        lg.log_phase(&cfg.run_name, current_k, step, flops_cum);
    }

    loop {
        if phase_step >= phase_total {
            if current_k >= cfg.k_max {
                break;
            }
            let old_k = current_k;
            current_k += 1;
            phase_step = 0;

            optim = AdamWConfig::new()
                .with_weight_decay(cfg.weight_decay as f32)
                .with_grad_clipping(Some(GradientClippingConfig::Norm(cfg.grad_clip as f32)))
                .init::<B, Snowman<B>>();

            if let Some(ref lg) = logger {
                lg.log_phase(&cfg.run_name, current_k, step, flops_cum);
            }
            if let Some(ref mut cb) = on_phase {
                cb(old_k, current_k, step, flops_cum);
            }
        }

        let lr = cosine_lr(cfg.lr, phase_step, phase_total);

        let (x, y) = train_data(device);

        // Random depth sampling: k ~ U(1..current_K)
        let k = (rand::random::<usize>() % current_k) + 1;

        let loss = model.forward_single(x, k, y);
        let ce = loss.clone().into_scalar().elem::<f64>();
        last_ce = ce;

        let grads = loss.backward();
        let grads_params = GradientsParams::from_grads(grads, &model);
        model = optim.step(lr, model, grads_params);

        let fpt = flop_single(k, p_emb, b_block, p_head);
        flops_cum += fpt as f64 * tokens_per_step as f64;

        if step % cfg.log_interval == 0 {
            let gates = model.gate_values();
            let gates_json = serde_json::to_string(&gates[..current_k]).unwrap();
            let pd_json = serde_json::to_string(&[ce]).unwrap();
            if let Some(ref lg) = logger {
                lg.log_step(&cfg.run_name, step, current_k, ce, lr, flops_cum, &gates_json, &pd_json);
            }
            if let Some(ref mut cb) = on_step {
                cb(step, current_k, ce, lr, flops_cum);
            }
        }

        if (step + 1) % cfg.eval_interval == 0 {
            if let Some(ref mut ed) = eval_data {
                let mut total = 0.0;
                for _ in 0..cfg.eval_batches {
                    let (x, y) = ed(device);
                    // Use autodiff model but detach the result (no grad tracking needed)
                    let logits = model.forward(x, current_k).detach();
                    let [b, t, v] = logits.dims();
                    let logits_flat = logits.reshape([b * t, v]);
                    let targets_flat = y.reshape([b * t]);
                    let loss = burn::nn::loss::CrossEntropyLossConfig::new()
                        .init(&logits_flat.device())
                        .forward(logits_flat, targets_flat);
                    total += loss.into_scalar().elem::<f64>();
                }
                let val_ce = total / cfg.eval_batches as f64;
                if val_ce < best_loss {
                    best_loss = val_ce;
                }
                if let Some(ref lg) = logger {
                    lg.log_eval(&cfg.run_name, step + 1, current_k, val_ce, flops_cum);
                }
                if let Some(ref mut cb) = on_eval {
                    cb(step + 1, current_k, val_ce, flops_cum);
                }
            }
        }

        step += 1;
        phase_step += 1;
    }

    let elapsed = t0.elapsed().as_secs_f64();

    let final_loss = if let Some(ref mut ed) = eval_data {
        let mut total = 0.0;
        for _ in 0..cfg.eval_batches {
            let (x, y) = ed(device);
            let logits = model.forward(x, cfg.k_max).detach();
            let [b, t, v] = logits.dims();
            let logits_flat = logits.reshape([b * t, v]);
            let targets_flat = y.reshape([b * t]);
            let loss = burn::nn::loss::CrossEntropyLossConfig::new()
                .init(&logits_flat.device())
                .forward(logits_flat, targets_flat);
            total += loss.into_scalar().elem::<f64>();
        }
        total / cfg.eval_batches as f64
    } else {
        last_ce
    };
    if final_loss < best_loss {
        best_loss = final_loss;
    }

    let result = TrainResult {
        best_loss,
        final_loss,
        total_steps: step,
        elapsed,
        total_flops: flops_cum,
        params: model.param_count(),
        block_params: b_block,
    };

    if let Some(ref lg) = logger {
        lg.log_summary(&cfg.run_name, "best_loss", &format!("{:.6}", result.best_loss));
        lg.log_summary(&cfg.run_name, "final_loss", &format!("{:.6}", result.final_loss));
        lg.log_summary(&cfg.run_name, "total_steps", &result.total_steps.to_string());
        lg.log_summary(&cfg.run_name, "elapsed", &format!("{:.1}", result.elapsed));
        lg.log_summary(&cfg.run_name, "total_flops", &format!("{:.3e}", result.total_flops));
        lg.log_summary(&cfg.run_name, "params", &result.params.to_string());
    }

    (model, result)
}

fn cosine_lr(base_lr: f64, phase_step: usize, phase_total: usize) -> f64 {
    base_lr * 0.5 * (1.0 + (std::f64::consts::PI * phase_step as f64 / phase_total as f64).cos())
}
